File: NewsExpress/app/src/main/java/com/example/newsexpress/retrofit/ApiService.java
package com.example.newsexpress.retrofit;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiService {

    private static final String BASE_URL = "https://newsapi.org/v2/everything/";
    private static final String API_KEY = "a8e9021dd36d403089f9d98ba7682fe7";
//
//    private final Retrofit retrofit = new Retrofit.Builder()
//            .addConverterFactory(GsonConverterFactory.create())
//            .baseUrl(BASE_URL)
//            .build();

//    public static Api apiMethod = null;
//
//    public static Api getApiMethod(){
//        if (apiMethod == null){
//            Retrofit retrofit = new Retrofit.Builder()
//                    .addConverterFactory(GsonConverterFactory.create())
//                    .baseUrl(BASE_URL)
//                    .build();
//
//            apiMethod = retrofit.create(Api.class);
//        }
//        return apiMethod;
//    }

//    Retrofit retrofit = new Retrofit.Builder()
//                    .addConverterFactory(GsonConverterFactory.create())
//                    .baseUrl(BASE_URL)
//                    .build();
//
//    Api api = retrofit.create(Api.class);


    private static ApiService instance = null;
    private Api myApi;

    private ApiService() {
        Retrofit retrofit = new Retrofit.Builder().baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        myApi = retrofit.create(Api.class);
    }

    public static synchronized ApiService getInstance() {
        if (instance == null) {
            instance = new ApiService();
        }
        return instance;
    }

    public Api getMyApi() {
        return myApi;
    }

}

File: NewsExpresss/app/src/main/java/com/example/newsexpresss/api/ApiService.kt
package com.example.newsexpresss.api

import com.example.newsexpresss.model.NewsReponse
import com.example.newsexpresss.utils.Constant.Companion.API_KEY
import com.example.newsexpresss.utils.Constant.Companion.BASE_URL
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import retrofit2.Call
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.http.GET
import retrofit2.http.Headers
import retrofit2.http.Query


private val moshi = Moshi.Builder()
        .add(KotlinJsonAdapterFactory())
        .build()

private val retrofit = Retrofit.Builder()
    .addConverterFactory(MoshiConverterFactory.create(moshi))
    .baseUrl(BASE_URL)
    .build()

object Api{
    val apiService: ApiService by lazy { retrofit.create(ApiService::class.java) }
}

interface ApiService {

    @GET("top-headlines")
    suspend fun getNewsList(
        @Query("apikey") key: String = API_KEY,
        @Query("country") country: String
    ): Response<NewsReponse>

}